SpherePacking/MagicFunction/b/psi.lean

```diff
@@ -51,65 +51,449 @@ def ψT' (z : ℂ) : ℂ := if hz : 0 < z.im then ψT ⟨z, hz⟩ else 0
+    SpecialLinearGroup.coe_GL_coe_matrix, SpecialLinearGroup.map_apply_coe,
+    RingHom.mapMatrix_apply, Int.coe_castRingHom, map_apply, of_apply, cons_val', cons_val_zero,
+    cons_val_fin_one, cons_val_one, Int.cast_one, ofReal_one, one_mul, Int.cast_zero,
+    ofReal_zero, add_zero, zpow_neg]
+
+private lemma slashS' (z : ℍ) (F : ℍ → ℂ) : ((F) ∣[(-2 : ℤ)] (S)) (z) =
+                         (F) (S • z) * (z : ℂ)^(2 : ℕ) := by
+  rw [SL_slash_apply, S, denom]
+  simp only [Int.reduceNeg, sl_moeb, coe2_smul, Fin.isValue,
+    SpecialLinearGroup.coe_GL_coe_matrix, SpecialLinearGroup.map_apply_coe,
+    RingHom.mapMatrix_apply, Int.coe_castRingHom, map_apply, of_apply, cons_val', cons_val_zero,
+    cons_val_fin_one, cons_val_one, Int.cast_one, ofReal_one, one_mul, Int.cast_zero,
+    ofReal_zero, add_zero, neg_neg, mul_eq_mul_left_iff]
+  have pow_coe_nat (a : ℂ) : a ^ (2 : ℕ) = a^(2 : ℤ) := by
+    rw [zpow_two, pow_two]
+  left
+  rw [pow_coe_nat]
+
+private lemma slashS'' (z : ℍ) (F : ℍ → ℂ): F (S • z) =
+                       (F ∣[(2 : ℤ)] (S)) (z) * (z : ℂ)^(2 : ℕ) := by
+  rw [slashS]
+  rw [mul_assoc]
+  simp only [sl_moeb, map_mul, Int.reduceNeg, zpow_neg]
+  have inv_mul_cancel (a : ℂ) (nonzero : a ≠ 0) : a⁻¹ * a = (1 : ℂ) := by
+    rw [mul_comm]
+    apply Complex.mul_inv_cancel
+    exact nonzero
+  have helper (a : ℂ) : a*a = 0 → a = 0 := by
+    simp only [mul_eq_zero, or_self, imp_self]
+  have pow_coe_nat (a : ℂ) : a ^ (2 : ℕ) = a^(2 : ℤ) := by
+    rw [zpow_two, pow_two]
+  have sp : (((z : ℂ) ^ (2 : ℤ))⁻¹ * (z : ℂ) ^ (2)) = 1 := by
+    apply inv_mul_cancel ((z : ℂ)^(2 : ℤ) )
+    field_simp
+    intro hP
+    apply z_nonzero z
+    apply helper
+    rw [← pow_two]
+    exact hP
+  rw [sp]
+  simp only [mul_one]
+
+private lemma slashT (z : ℍ) (F : ℍ → ℂ) : ((F) ∣[(2 : ℤ)] (T)) (z) = (F) (T • z) := by
+  rw [SL_slash_apply, T, denom]
+  simp only [sl_moeb, coe2_smul, Fin.isValue, SpecialLinearGroup.coe_GL_coe_matrix,
+    SpecialLinearGroup.map_apply_coe, RingHom.mapMatrix_apply, Int.coe_castRingHom, map_apply,
+    of_apply, cons_val', cons_val_zero, cons_val_fin_one, cons_val_one, Int.cast_zero,
+    ofReal_zero, zero_mul, Int.cast_one, ofReal_one, zero_add, Int.reduceNeg, zpow_neg,
+    one_zpow, inv_one, mul_one]
+
+private lemma slashT' (z : ℍ) (F : ℍ → ℂ) : ((F) ∣[(-2 : ℤ)] (T)) (z) =  (F) (T • z) := by
+  rw [SL_slash_apply, T, denom]
+  simp only [sl_moeb, coe2_smul, Fin.isValue, SpecialLinearGroup.coe_GL_coe_matrix,
+    SpecialLinearGroup.map_apply_coe, RingHom.mapMatrix_apply, Int.coe_castRingHom, map_apply,
+    of_apply, cons_val', cons_val_zero, cons_val_fin_one, cons_val_one, Int.cast_zero,
+    ofReal_zero, zero_mul, Int.cast_one, ofReal_one, zero_add, Int.reduceNeg, neg_neg, one_zpow,
+    mul_one]
+ -- no need for slashT'', as ← slashT already fulfils that role
+
+private lemma S_mul_T : S * T = ⟨!![0, -1; 1, 1], by norm_num [det_fin_two_of]⟩ := by
+  ext (i : Fin 2) (j : Fin 2)
+  fin_cases i; fin_cases j
```

Comment:
You can golf this by adding `<;>` after `fin_cases j`. Then only the first `simp [mul_apply, S, T]` is needed: you can remove the other two.
(`<;>` is a nice bit of syntax that applies whatever immediately follows it to all of the goals in the proof. It's great for golfing, and it's definitely something worth using in the future.)